docs/DesignDocs/CustomAttestation.md

```diff
@@ -0,0 +1,754 @@
+users who want to specify their endorsements (information from a second source
+used for verification), instead of using the set of endorsements provided by Open
+Enclave.
+
+Overall, there has been interest in enhancing Open Enclave's APIs to support
+custom attestation formats to enable these scenarios.
+
+Terminology
+-----------
+
+This document uses the following terminology defined below. Note that
+these definitions are consistent with the terms defined in the
+[Remote Attestation Procedures (RATS)](https://datatracker.ietf.org/wg/rats/about/)
+working group.
+
+- Claims
+  - Claims are statements about a particular subject. They consist of
+    name-value pairs containing the claim name, which is a string, and
+    claim value, which is arbitrary data. Example of claims could be
+    [name="version", value=1] or [name="enclave_id", value=1111].
+- Evidence
+  - This is the data about the enclave that is produced and signed by it.
+    The SGX report would be an example of evidence.
+- Endorsements
+  - This is additional data that used in the evidence verification process,
+    but is not produced by the enclave. An example of an endorsement would be
+    the quoting enclave's identity used in SGX remote attestation, because it
+    is retrieved from Intel's servers, rather than the enclave.
+- Verifier
+  - The verifier is responsible for taking in the evidence and endorsements
+    and deciding if the enclave is trustworthy.
+- Relying party
+  - The relying party is the entity interested in communicating with an
+    enclave. The enclave must attest to the relying party before the
+    relying party can trust it. The relying party can also play the role
+    of the verifier, but it does not necessarily have to.
+
+Specification
+-------------
+
+To support custom attestation formats, this document proposes adding a plugin
+model for attestation. The Open Enclave SDK will define a set of common APIs
+that each plugin must implement. Each plugin will define an UUID to distinguish
+it from other plugins.
+
+Futhermore, there will be additional attestation "plugin aware" APIs that are
+analogous to `oe_get_report` and `oe_verify_report` called `oe_get_evidence`
+and `oe_verify_evidence` respectively. These align with the ones proposed in
+the [endorsements design doc](RemoteAttestatonCollaterals.md). There will also
+be functions for registering and unregistering plugins called
+`oe_register_attestation_plugin` and `oe_unregister_attestation_plugin`. The user
+can statically link in their desired plugin and call the register plugin function.
+The attestation data can be retrieved from the "plugin aware" analogue of
+`oe_get_report` with the desired UUID. The generated data will have the UUID in
+its header. The user can call the analogue of `oe_verify_report` to verify the
+data and the Open Enclave runtime can use this UUID to determine what plugin
+verification routine to run.
+
```

Comment:
Explain whether the plugin is loaded into, and called by, the host side or the enclave side of the app.

Reply:
Added an explanation.